File: killrvideo-service-comments/src/main/java/com/killrvideo/service/comment/dto/Comment.java
package com.killrvideo.service.comment.dto;

import com.datastax.oss.driver.api.mapper.annotations.ClusteringColumn;
import com.datastax.oss.driver.api.mapper.annotations.Computed;
import com.killrvideo.dse.dto.AbstractEntity;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.validator.constraints.Length;

import javax.validation.constraints.NotNull;
import java.time.Instant;
import java.util.UUID;

/**
 * Bean standing for comment on video.
 *
 * @author DataStax Developer Advocates team.
 */
@Getter @Setter @NoArgsConstructor
public class Comment extends AbstractEntity {
    private static final long serialVersionUID = 7675521710612951368L;

    /**
     * Column names in the DB.
     */
    public static final String COLUMN_USERID = "userid";
    public static final String COLUMN_VIDEOID = "videoid";
    public static final String COLUMN_COMMENTID = "commentid";
    public static final String COLUMN_COMMENT = "comment";

    @NotNull
    protected UUID userid;

    @NotNull
    protected UUID videoid;

    @NotNull
    @ClusteringColumn
    protected UUID commentid;

    @Length(min = 1, message = "The comment must not be empty")
    protected String comment;

    @NotNull
    @Computed("toTimestamp(commentid)")
    private Instant dateOfComment;

    /**
     * Copy constructor.
     *
     * @param other comment.
     */
    public Comment(Comment other) {
        this.commentid = other.commentid;
        this.userid = other.userid;
        this.videoid = other.videoid;
        this.comment = other.comment;
        this.dateOfComment = other.dateOfComment;
    }
}

File: killrvideo-service-comments/src/main/kotlin/com/killrvideo/service/comment/dto/Comment.kt
package com.killrvideo.service.comment.dto

import java.time.Instant
import java.util.*
import javax.validation.constraints.NotNull
import javax.validation.constraints.Size

/**
 * Bean standing for comment on video.
 *
 * @author DataStax Developer Advocates team.
 */
data class Comment(
    var userid: @NotNull UUID?,
    var videoid: @NotNull UUID?,
    var commentid: @NotNull UUID?,
    var comment: @Size(min = 1, message = "The comment must not be empty") String?,
    var dateOfComment: @NotNull Instant?
) {
    /**
     * Copy constructor.
     *
     * @param other comment.
     */
    constructor(other: Comment) : this(
        userid = other.userid,
        videoid = other.videoid,
        commentid = other.commentid,
        comment = other.comment,
        dateOfComment = other.dateOfComment
    )

    companion object {
        /**
         * Column names in the DB.
         */
        const val COLUMN_USERID = "userid"
        const val COLUMN_VIDEOID = "videoid"
        const val COLUMN_COMMENTID = "commentid"
        const val COLUMN_COMMENT = "comment"
        const val COLUMN_COMMENT_TIMESTAMP = "comment_timestamp"
    }
}
